crates/matrix-sdk-crypto/src/machine/tests/decryption_verification_state.rs

```diff
@@ -373,3 +381,254 @@ async fn test_verification_states_multiple_device() {
+            .clone(),
+        legacy_session: false,
+    };
+    bob.store().save_inbound_group_sessions(&[session.clone()]).await.unwrap();
+
+    check_decryption_trust_requirement(
+        &bob,
+        &event,
+        room_id,
+        &[
+            (TrustRequirement::Untrusted, true),
+            (TrustRequirement::CrossSignedOrLegacy, false),
+            (TrustRequirement::CrossSigned, false),
+        ],
+    )
+    .await;
+
+    // Bob later gets Alice's identity and cross-signed device keys
+    bob.bootstrap_cross_signing(false).await.unwrap();
+    let cross_signing_requests = alice.bootstrap_cross_signing(false).await.unwrap();
+    let upload_signing_keys_req = cross_signing_requests.upload_signing_keys_req;
+    let alice_msk: MasterPubkey = upload_signing_keys_req.master_key.unwrap().try_into().unwrap();
+    let alice_ssk: SelfSigningPubkey =
+        upload_signing_keys_req.self_signing_key.unwrap().try_into().unwrap();
+    let upload_keys_req = cross_signing_requests.upload_keys_req.unwrap().clone();
+    assert_let!(
+        OutgoingRequests::KeysUpload(device_upload_request) = upload_keys_req.request.as_ref()
+    );
+    bob.store()
+        .save_device_data(&[DeviceData::try_from(
+            &device_upload_request
+                .device_keys
+                .as_ref()
+                .unwrap()
+                .deserialize_as::<DeviceKeys>()
+                .unwrap(),
+        )
+        .unwrap()])
+        .await
+        .unwrap();
+    bob.store()
+        .save_changes(Changes {
+            identities: IdentityChanges {
+                new: vec![OtherUserIdentityData::new(alice_msk.clone(), alice_ssk.clone())
+                    .unwrap()
+                    .into()],
+                ..Default::default()
+            },
+            ..Default::default()
+        })
+        .await
+        .unwrap();
+
+    // Since the sending device is now cross-signed by Alice, it should be
+    // decryptable in all modes
+    check_decryption_trust_requirement(
+        &bob,
+        &event,
+        room_id,
+        &[
+            (TrustRequirement::Untrusted, true),
+            (TrustRequirement::CrossSignedOrLegacy, true),
+            (TrustRequirement::CrossSigned, true),
+        ],
+    )
+    .await;
+}
+
+#[async_test]
+async fn test_decryption_trust_with_identity_change() {
+    // Test that the trust requirement is correctly handled when a user's
+    // cross-signing identity changes.
+    let (alice, bob) = get_machine_pair_with_setup_sessions_test_helper(
+        tests::alice_id(),
+        tests::user_id(),
+        false,
+    )
+    .await;
+    bob.bootstrap_cross_signing(false).await.unwrap();
+    let room_id = room_id!("!test:example.org");
+    let (event, session_id) = encrypt_message(&alice, room_id, &bob, "Secret message").await;
+
+    // Simulate Alice's cross-signing key changing after having been verified by
+    // setting the `previously_verified` flag
```

Comment:
we don't actually seem to set the `previously_verified` flag until about 50 lines later?